Refactor to use shape arguments

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #64323 Add embedding shape analysis
* #64121 Max Pool with indices
* #63530 Add Maxpool to shape analysis / Opinfo
* #64300 Add symbolic shape comparison optimization
* **#64299 Refactor to use shape arguments**
* #63516 Add view with negative dim
* #63615 Generalize expand logic
* #63407 Add permute, arange
* #63365 Add support for slice, selec twith int, index_select
* #63099 Add squeeze, unsqueeze, transpose shape functins
* #63050 Add batch of unary functions

Differential Revision: [D30738141](https://our.internmc.facebook.com/intern/diff/D30738141)

Commit:
Refactor to use shape arguments

[ghstack-poisoned]

Krovatkin commented on torch/csrc/jit/passes/symbolic_shape_analysis.cpp:
```diff
@@ -43,6 +43,91 @@ static bool symbolic_shape_analysis_test_mode = false;
+};
+
+struct ShapeArguments {
+  using MaybeSymbolicShapeOrInt =
```
we don't seem to need it?

Commit:
Update on "Refactor to use shape arguments"




[ghstack-poisoned]

Krovatkin commented on torch/csrc/jit/passes/symbolic_shape_analysis.cpp:
```diff
@@ -43,6 +43,91 @@ static bool symbolic_shape_analysis_test_mode = false;
+// - A Constant Integer
+// - Neither of the above. The third case can occur due to inputs to
+// ops like view that accept negative values. Maintaining the distinction
+// between an unknown symbolic dimension and an unknown integer allows
```
unkown symbolic dimension is just a fresh symbolic dimension?

eellison replied:
yes, that's correct. maybe `unknown dimension`

Krovatkin commented on torch/csrc/jit/passes/symbolic_shape_analysis.cpp:
```diff
@@ -43,6 +43,91 @@ static bool symbolic_shape_analysis_test_mode = false;
+// ops like view that accept negative values. Maintaining the distinction
+// between an unknown symbolic dimension and an unknown integer allows
+// us to optimize out comparisons to values < 0 (symbolic shapes are always >=
+// 0) For example, a call like graph(%y: Tensor(SS(-1), 10, 10), %inp: int):
```
where does a comparison happen in this example? I'm a little bit unclear how symbolic shapes are always >= 0 when they are backed up by negative integers?

eellison replied:
symbolic shapes represent dimension values. You cannot create a dimension < 0, it would be an error. That they are stored as negative integers is an implementation detail as you said.

Krovatkin replied:
got it! yes, a symbolic dimension should still be a positive number since it's supposed to stand for some concrete dimension value.

Krovatkin reviewed (commented):
I think I don't understand this PR fully. I thought the issue was that negative numbers can be interpreted as either symbolic dimensions or negative indices, but the comment for ShapeArg seems to say otherwise.

Commit:
Update on "Refactor to use shape arguments"




[ghstack-poisoned]

eellison commented on torch/csrc/jit/passes/symbolic_shape_analysis.cpp:
```diff
@@ -43,6 +43,91 @@ static bool symbolic_shape_analysis_test_mode = false;
+    this->second = int_value;
+  }
+
+  ShapeArg(c10::ShapeSymbol ss) {
```
IMO c10::variant is more confusing than just have a pair...

eellison reviewed (commented):
> I think I don't understand this PR fully. I thought the issue was that negative numbers can be interpreted as either symbolic dimensions or negative indices, but the comment for ShapeArg seems to say otherwise.

The motivation for the PR is to maintain the difference between a symbolic dimension, a constant number (positive or negative), and a general unknown integer (which can be positive or negative).

You can't use a constant negative number for a shape because it will be interpreted as a symbolic dimension, and we want to distinguish the two. It might help to look at the subsequent PR that uses the information we're maintaining here in an optimization.

Krovatkin commented on torch/csrc/jit/passes/integer_value_refinement.cpp:
```diff
@@ -89,8 +89,10 @@ struct IntegerValueRefiner {
-            constant_as<int64_t>(other_if_block->outputs().at(i));
+        auto other_output = other_if_block->outputs().at(i);
+        auto other_const_value = other_output->type()->cast<IntType>()
+            ? constant_as<int64_t>(other_output)
```
wouldn't constant_as<**int64_t**> return a value iff the type is `IntType`?

eellison replied:
Unfortunately not, it throws an error if you run constant_as<int64_t> on a None constant. I actually had a PR to fix this and it was rejected lol.

Krovatkin commented on torch/csrc/jit/passes/symbolic_shape_analysis.cpp:
```diff
@@ -31,7 +31,7 @@ XXX: this is still in prototype phase and has much work left to do, including
 - Refactor APIs
 - Add decent coverage of common ops
-- Add shape analysis pass on Graph that handles Ifs and Loops
+- Add shape analysis pass on Graph that handles Loops
```
you don't seem to be removing if handling in this PR. why change this comment?

eellison replied:
It was done already and i forgot to remove this line

Krovatkin reviewed (approved):
:shipit: looks good. Initially I got this PR backwards.

Commit:
Update on "Refactor to use shape arguments"




[ghstack-poisoned]

Commit:
Update on "Refactor to use shape arguments"




[ghstack-poisoned]

Commit:
Update on "Refactor to use shape arguments"




[ghstack-poisoned]

Commit:
Update on "Refactor to use shape arguments"




[ghstack-poisoned]

Commit:
Update on "Refactor to use shape arguments"




[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30738141).

Commit:
Update on "Refactor to use shape arguments"


Differential Revision: [D30738141](https://our.internmc.facebook.com/intern/diff/D30738141)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30738141).

Commit:
Update on "Refactor to use shape arguments"


Differential Revision: [D30738141](https://our.internmc.facebook.com/intern/diff/D30738141)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30738141).
